functionary/prompt_template/prompt_template_v2.py

```diff
@@ -0,0 +1,337 @@
+            recipient = response[
+                recipient_index + len(self.recipient_token) : content_index
+            ].strip()
+            content = response[content_index + len(self.content_token) :].strip()
+            # print(f"recipient: {recipient}, content={content}")
+            if recipient == "all":
+                text_response = content
+            else:
+                tool_calls.append(
+                    {
+                        "function": {"name": recipient, "arguments": content},
+                        "id": get_random_tool_call_id(),
+                        "type": "function",
+                    }
+                )
+
+        return {"role": "assistant", "content": text_response, "tool_calls": tool_calls}
+
+    def pre_process_messages_before_inference(self, messages: List[Dict]) -> List[Dict]:
+        """re-order the messages where role = tool to match the order in tool_calls by tool_call_id
+        Args:
+            messages (List[Dict]): list of messages containing: tool_call_id
+
+        Returns:
+            List[Dict]: _description_
+        """
+        result = []
+        index = 0
+        while index < len(messages):
+            message = messages[index]
+            tool_calls = message.get("tool_calls", None)
+
+            result.append(message)
+            if message["role"] == "assistant" and tool_calls:
+                num_calls = len(tool_calls)
+                if (
+                    tool_calls[0].get("id", None) is not None
+                ):  # if tool_call contains "id" for mapping
+                    tool_call_ids = [item["id"] for item in tool_calls]
+
+                    tool_messages = [messages[index + 1 + j] for j in range(num_calls)]
+                    id_2_tool_messages = {
+                        item["tool_call_id"]: item for item in tool_messages
+                    }
+                    new_messages = [id_2_tool_messages[cid] for cid in tool_call_ids]
+
+                    result.extend(new_messages)
+                    index += num_calls + 1
+                else:
+                    index += 1
+            else:
+                index += 1
+        return result
+
+    def get_function_delta_response(
+        self,
+        current_state: Dict,
+        delta_text: str,
+        first_call: bool,
+        return_role: bool,
+        finish_reason: Optional[str],
+    ) -> Dict:
+        """Return delta for tool_call in streaming
+
+        Args:
+            current_state (Dict): _description_
+            delta_text (str): _description_
+            first_call (bool): _description_
+            return_role (bool): _description_
+            finish_reason (Optional[str]): _description_
+
+        Returns:
+            Dict: _description_
+        """
+        return {
+            "delta": {
+                "content": None,
+                "function_call": None,
+                "role": None if not return_role else "assistant",
```

Comment:
Why do we need to return None for "role"?

Reply:
because at the first time role=assistant and later, role is always None
This is the format from openAI streaming, I just tried to replicate exactly

Reply:
Ah ok